lib: nrf_modem: enable traces using AT command if `_TRACE_ENABLED`

lemrey opened the pull request:
Let nrf_modem_lib_init() enable trace output if
CONFIG_NRF_MODEM_LIB_TRACE_ENABLED is set.

This was done in the Link Controller or elsewhere,
I think it's a good idea to have it here instead.

Introduce CONFIG_NRF_MODEM_LIB_TRACE_LEVEL to
control the trace level to be set upon initialization.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-7328
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

NordicBuilder commented:
### Integration test specification

| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-ci-nrfconnect-boot-fw-update | X | | |
| test-fw-nrfconnect-nrf-iot_libmodem | X | | |
| test-fw-nrfconnect-nrf-iot_lwm2m | X | | |
| test-fw-nrfconnect-nrf-iot_mosh | X | | |
| test-fw-nrfconnect-nrf-iot_positioning | X | | |
| test-fw-nrfconnect-nrf-iot_samples | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sub/job/test-fw-nrfconnect-nrf/job/master/22411/artifact/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

balaji-nordic commented on boards/arm/thingy91_nrf9160/board_nonsecure.c:
```diff
@@ -17,7 +17,6 @@ LOG_MODULE_REGISTER(board_nonsecure, CONFIG_BOARD_LOG_LEVEL);
 				"2,1710,2200,3,824,894,4,880,960,5,791,849," \
 				"7,1565,1586"
 #define AT_CMD_COEX0		"AT%%XCOEX0=1,1,1565,1586"
-#define AT_CMD_TRACE		"AT%%XMODEMTRACE=0"
```
I tried to figure this out too but couldn't. I ended up with this commit https://github.com/nrfconnect/sdk-nrf/pull/702/files#diff-fc089568a197727192e08ec349a493d33aaa4f5516f537faaadd40ef91fa975cR37 where they disable the traces before sending the antenna configuration. But there is no reasoning mentioned. Perhaps this was added by mistake. And I believe this is the root cause for CIA-606 where the traces stopped working for thing91 after https://github.com/nrfconnect/sdk-nrf/commit/305d1e40dbd449e310b18b74e142f0591fb8a278 . I believe the on_modem_init enables traces but this code here disables traces immediately.

balaji-nordic replied:
@lemrey  Any reason why you had to do this change as part of this PR?

lemrey replied:
We are trying to set the default trace level in one place in the application. It seems here it was disabled, and then again re-enabled during `lte_lc`  initialization. When the piece of code in `lte_lc` that enabled trace was moved to `lte_lc_hooks` it probably ran before `board.c` and the trace was disabled afterwards.
@jorgenmk can you check if that issue is gone in this PR?

jorgenmk replied:
@lemrey I'll check

jorgenmk replied:
@lemrey It appears we do have traces on thingy91

jtguggedal replied:
I think disabling the traces is a relic from the time Thingy:91's USB bridge didn't support tracing

balaji-nordic reviewed (approved):
The changes look good. Just one clarification needed.

lemrey commented on lib/nrf_modem_lib/Kconfig.modemlib:
```diff
@@ -22,28 +22,51 @@ config NRF_MODEM_LIB_TRACE_ENABLED
+	  When enabled, a portion of RAM (called Trace region) will be shared with the modem to receive modem's trace data.
+	  The size of the Trace region is defined by the NRF_MODEM_LIB_SHMEM_TRACE_SIZE option.
+	  Trace data is output on the chosen trace medium.
+	  To stop traces or to activate a smaller subset of traces, use the AT command %XMODEMTRACE or the nrf_modem_lib_trace API.
```
Thanks, I clarified this.

simensrostad commented on lib/nrf_modem_lib/nrf_modem_lib.c:
```diff
@@ -130,6 +134,13 @@ static int _nrf_modem_lib_init(const struct device *unused)
 	k_mutex_unlock(&slist_mutex);
 
+#if defined(CONFIG_NRF_MODEM_LIB_TRACE_LEVEL)
+	err = nrf_modem_lib_trace_start(CONFIG_NRF_MODEM_LIB_TRACE_LEVEL);
```
```suggestion
	int err = nrf_modem_lib_trace_start(CONFIG_NRF_MODEM_LIB_TRACE_LEVEL);

```

simensrostad replied:
nit: Looks like `err` is only used here. May I suggest declaring it here as well to save a few bytes when `CONFIG_NRF_MODEM_LIB_TRACE_LEVEL` is disabled.

simensrostad replied:
Hmm I guess it will be optimized away anyway so no big deal.

lemrey commented:
I have removed the entire `board_nonsecure.c` file, since it contained only trace and antenna related setup, which are now moved to `nrf_modem_lib` and `modem_antenna` respectively.

gregersrygg reviewed (approved):
LGTM 👍

lemrey commented:
Rebased yet again

mbolivar-nordic commented:
> Rebased yet again

This needs another rebase :(

lemrey commented:
Let's see now

lemrey commented:
And rebased
